src/metabase/upload.clj

```diff
@@ -190,15 +243,21 @@
 
 (mu/defn ^:private settings->type->check :- type->check-schema
   [{:keys [number-separators] :as _settings}]
-  {::boolean-or-int  boolean-or-int-string?
-   ::boolean         boolean-string?
-   ::offset-datetime offset-datetime-string?
-   ::date            date-string?
-   ::datetime        datetime-string?
-   ::int             (regex-matcher (int-regex number-separators))
-   ::float           (regex-matcher (float-regex number-separators))
-   ::varchar-255     varchar-255?
-   ::text            (constantly true)})
+  (let [explicit-int? (regex-matcher (int-regex number-separators))
+        float-or-int? (regex-matcher (float-or-int-regex number-separators))]
+    {::*boolean-or-int* boolean-or-int-string?
+     ::boolean          boolean-string?
+     ::offset-datetime  offset-datetime-string?
+     ::date             date-string?
+     ::datetime         datetime-string?
+     ;; Note that this violates a desirable property - that the type-check for every ancestor of the value's type
+     ;; will pass the check as well. Unfortunately we have to break this property for the way that type-coercion
+     ;; abused the type hierarchy.
+     ::int              explicit-int?
```

Comment:
I'm probably missing something obvious, but can you give an example that breaks this property? because it's not obvious to me maybe you should leave the example in a comment

Reply:
Oh, `float-or-int-regex` doesn't match "1". That surprised me.

Reply:
Revenge of the union / intersection ambiguity.

You're right that actually this works out a bit better if we relax the regex, that penny only dropped after I wrote the comment. 

Against changing, it would be treacherous to have  different semantics for `boolean / int` versus `float / int`.

Reply:
I found an existing bug, which extends to this PR. I think the bug was introduced when we started reducing over ancestors of the current type instead of starting from the bottom of the graph every time.

Here's a passing test
```
(deftest load-from-csv-int-and-float-test
  (testing "Upload a CSV file with integers and floats in the same column"
    (mt/test-drivers (mt/normal-drivers-with-feature :uploads)
      (with-mysql-local-infile-on-and-off
        (with-upload-table!
          [table (let [table-name (mt/random-name)]
                   (@#'upload/load-from-csv!
                    driver/*driver*
                    (mt/id)
                    table-name
                    (csv-file-with ["float,oops"
                                    "1,1.0" ;; <--- these columns just differ in order
                                    "1.0,1"])) 
                   (sync-upload-test-table! :database (mt/db) :table-name table-name))]
          (testing "Check the data was uploaded into the table correctly"
            (is (= [[1 1.0 "1.0"]
                    [2 1.0 "1"]] ;; <---- these columns have different types
                   (rows-for-table table)))))))))
```

to fix the bug, we need to update `float-regex` to satisfy the property in your comment, and the `float-or-int?` function in this PR too (or come up with another model)

Reply:
Well spotted. Ironically if I recall the ancestor walk was actually introduced to fix exactly this issue, the least common ancestor of `boolean-or-int` and `float` being `varchar-255`. Sticking to the pure semantics with `::explicit-int` kept with the spirit of the topological sort, so

I'll take a look at relaxing the regex, and I guess the name wrangling that entails.

Reply:
You'll never believe the naming I ended up on 🫣